File: 2024/day_7/solution.ex
defmodule Day7 do
  def parse(line) do
    [ans, nums] = String.split(line, ":")
    {String.to_integer(ans), nums |> String.trim() |> String.split() |> Enum.map(fn n -> String.to_integer(n) end)}
  end

  def dfs([], total, ans) do
    total == ans
  end

  def dfs([next | remaining], total, ans) do
    mul = fn n, t -> t * n end
    add = fn n, t -> t + n end
    concat = fn n, t -> String.to_integer(Integer.to_string(t) <> Integer.to_string(n)) end

    [mul, add, concat]
      |> Enum.map(fn op -> dfs(remaining, op.(next, total), ans) end)
      |> Enum.reduce(false, fn v, acc -> v or acc end)
  end

  def isPossible({ans, nums}) do
    [first | remaining] = nums
    dfs(remaining, first, ans)
  end

  def run(file) do
    File.stream!(file)
      |> Stream.map(fn line -> parse(line) end)
      |> Stream.filter(fn v -> isPossible(v) end)
      |> Stream.map(fn {ans, _} -> ans end)
      |> Enum.reduce(0, fn ans, acc -> ans + acc end)
  end
end

File: 2024/day_7/solution.py
import sys 

def parse(l):
    ans, nums = l.split(":")
    return int(ans), [int(x) for x in nums.strip().split(" ")]

def isPossible(ans, nums):
    def add(i, t):
        return nums[i] + t

    def mul(i, t):
        return nums[i] * t

    def concat(i, t):
        return int(f"{t}{nums[i]}")

    def dfs(i, t):
        if i == len(nums) and t == ans:
            return True
        if i == len(nums):
            return False
        for op in [add, mul, concat]:
            if dfs(i + 1, op(i, t)):
                return True
        return False

    return dfs(1, nums[0])

if __name__ == "__main__":
    file = sys.argv[1]
    ops = []
    with open(file) as f:
        for l in f:
            ans, nums = parse(l)
            ops.append((ans, nums))

    res = 0
    for ans, nums in ops:
        if isPossible(ans, nums):
            res += ans
    print(res)